modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+
+----
+toUser(123, "John")
+toUser("123", true)
+----
+
+Though you might expect the second call to fail, autocoercion enables the call to be successful, and both calls work. In the case of the second function call, there is an autocoercion of the `String` value `"123"` to the `Number` value `123`, and the `Boolean` value `true` to the `String` value `"true"`, which makes it possible to call the function. +
+An example of a call that throws a type checking error is `toUser("a 12", "John")` because the String `"a 12"` isn't coercible to a `Number` type. +
+
+DataWeave also uses a global type inference algorithm to validate your code even if no types are specified as
+constraints.
+
+[NOTE]
+ The use of constraints is optional, but constraints can be useful in big scripts or multiple scripts. The type system can help you avoid bugs in your DataWeave logic because it causes the type-checking algorithm to run.
+
+The types supported by the DataWeave type system are divided into three categories:
+
+* Simple Types
+* Composite Types
+* Complex Types
+
+== Simple Types
+
+Simple types represent values such as strings, Booleans, and so on. These values are atomic; they are not composed of other values. These are the simple types:
+
+* `String`
+* `Boolean`
+* `Number`
+* `Regex`
+* `Null`
+* Temporal: `Date`, `DateTime`, `LocalDateTime`, `LocalTime`, `Time`, `Period`
+//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType
+
+=== Null
+
+`Null` is a type of only one value, the value `null`, which means that `null` cannot be assigned to the type `String` or any other type except for the type `Null`.
+
+For example, the `repeat` function in the String module has the signature `repeat(String, Number): String`, which means that the function only accepts as a first parameter a value of type `String`, and as a second parameter a value of type `Number.` The function returns a value of type `String`.
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat("a", 3)
+----
+
+This returns `&quot;aaa&quot;`, but:
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(null, 3)
+----
+
+Throws an error because the value `null` cannot be assigned to the parameter that expects a value of type String.
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(Null, 3)
+----
+
+The script assigns to the first parameter a value that is the type `Null` (the type of this value
+is `Type&lt;Null&gt;`), and it returns `&quot;NullNullNull&quot;` because there is an autocoercion that coerces the value to the type `String` (`&quot;Null&quot;`).
+
+Some other functions have an overloaded signature for the function (the function has multiples definitions for different parameter types) that let you call it with `null`.
```

Comment:
Is this still correct?
```suggestion
Some functions have an overloaded signature (the function has multiple definitions for different parameter types) that let you call it with `null`.
```
Do you call with `null` or `Null`? All this type vs. value is starting to cloud my mind....